### PR TITLE
Responsive Chart, Additional Features and, Refactoring and organizing.

### PR DESCRIPTION
**Pull request by ronram5126**
### Responsive Chart
Instead of having a single svg for both graph and legend, only graph is mostly in svg. The legend is created with React Div. This is done to utilize the responsive css positioning of graph and legend. Css flex has been used to achieve responsiveness. The flex-wrap is set to wrap by default which means if the chart don't have enough space the graph will put chart on top and graph at bottom.

There are additional property to set positioning of the legend. This also removes the complexity related to placement of LegendItem.

- **legendSide** : Determines position of legend relative to graph. By default legend is placed on right. Possible values are:
   - top
   - left
   - right
   - bottom
- **verticalAlign**: Determines how the legend is vertically aligned with respect to graph. By default legend is placed in the middle: Possible values are:
   - top
   - middle
   - bottom
 - **horizontalAlign**: Determines how the legend is horizontally aligned with respect to graph. By default legend is placed in the left: Possible values are:
   - left
   - center
   - right

## Additional Features
- Stroke for the chart could be set to item's color. This will set the stroke color for item to the fill colors. The default stroke color has been modified to use this feature as it looks more modern.
- ChartSize replaces width and height. An additional gain from using this format is that legend can be sized dynamically. This means that the Graph will behave as the square svg image. The ChartSize determines the size of this square svg.  
- ```labelRenderer``` property has been added. This is a function to convert label and item into the display that is shown in legend. The function should take label string and item object as the argument and return React Element or string. The availability of the property and its calling is done under **LegendItem**. This extends the customization of the program.

## Refactoring and organizing
- The default colors have been moved to colors.ts so its easier to extend and updated in future.
- As majority of property under context is just being used by just ArcPath, and the property isn't required in nested setup, there is no requirement for declaring Context. it just introduces unwanted complexity and potential side-effects.
- As we are using the props to instead of context the Type of Property will inherit from the base property (just the way context was declared using the base property ). Therefore it is better to put them all in one file. This also is a good practice as it creates more hierarchical file dependency i.e. everyone depends on interface.ts instead of everyone is depending on each other.

PS: Great Repo

### CODE REVIEW

File: src/lib/Color.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const colors =[
```
Comment (vonbearshark) [inline src/lib/Color.ts]: It seems like these aren't exactly the original colors, and there are some dupes

File: src/lib/LegendItem.tsx [hunk 2]
```diff
@@ -15,28 +16,23 @@ const LegendItem: React.FC<Props> = ({ item }) => {
...
+    >      
+      <div className={`${className}-${classSuffix}-label ${classNames}`}>
+        <svg
+          width="0.89em"
+          height="0.89em"
+          viewBox="0 0 100 100" 
+          style={{ marginRight: "0.5em" }}
+        ><rect {...restItemRenderProps} width={100} height={100} />
+        </svg>
+        {typeof labelRenderer === "function"? labelRenderer(label, item) : `${label} - ${value}`}
```
Comment (vonbearshark) [inline src/lib/LegendItem.tsx]:
```suggestion
        {labelRenderer ? labelRenderer(label, item) : `${label} - ${value}`}
```

File: src/index.css [hunk 3]
```diff
@@ -31,18 +31,27 @@ body {
   width: 750;
 }
 
+.donutchart-container {
+  width: max-content;
+  gap: 2em!important;
+}
+.donutchart-graph {
```
Comment (vonbearshark) [inline src/index.css]: Why did these need to be updated?

**Review by vonbearshark — CHANGES_REQUESTED** [review]
Hey @ronram5126 . Thanks for taking the time. I am not totally sold on these changes
- I am open to splitting the legend into a separate div. But I'm concerned about that legend then having an embedded SVG. I'm worried about the SVG styling within that legend div. It seems like it would be tough to reconcile flex and the svg styling, but I could be wrong. And it seems like a little too much complexity, the style arrangements. But again, could be wrong there too
- What is the value to the chart behaving as a square, versus the flexibility in a non-square?
- What's the value in hierarchical file dependency?

Generally, I'm concerned about landing these changes and breaking existing usages. The extent of these changes seems to warrant a new version. I think there's otherwise opportunity for existing styles to break

Meanwhile users _could_ replicate this right now, right? By turning the legend off and creating their own non-SVG legend? Generally, I try to do _less_ with this library and push towards that kind of solution

At any rate, do you have some visuals of the changes?

File: src/lib/Color.ts (hunk 1 above)
Comment (ronram5126) [reply src/lib/Color.ts]: I have removed the duplicate color. hadn't check it sorry.

File: src/index.css (hunk 3 above)
Comment (ronram5126) [reply src/index.css]: Those were there while I was testing preview. I've removed them and updated default style for donutchart-container itself.

**Comment by ronram5126** [comment]
Hi @vonbearshark, 

- I've removed the SVG from LegendItem instead am using span with style now. 
- The graph svg is more like circle so it is better if we keep the chart as square. The height of the svg will always be almost equal to width. so instead of having two different inputs which will contain same data we have a single size input. 
    - The height also creates issue when there is responsive view. a non square chart means that the chart is contained within a bigger box causing unwanted spacing.
- I'm not 100% sure what you mean by hierarchical file dependency. 
    - If you are asking why supply props instead of using context. 
        - Context is mostly used in scenarios where you cannot directly inject the property or same property is being used in deep nested children.
        - In this use case only ArcPath uses most of the context property and some are getting used by LegendItem which are both direct  accessible children. Therefore it is better to use the property to pass the required values.
        - Using property over context reduces the complexity and improves the performance. It also reduces amount of hooks used to get the context. 
     - If you mean with having file hierarchy for import
        - Its actually is a old good practice, not essential with modern development tools but prevents issues with testing and some plugins that might not be able to resolve circular dependencies (most stuff does).
 

I do agree with you that this is a radical change an update to versioning and doc is required. But I'm not sure how to approach this as I don't know what your approach is and I'm kind of busy. I needed a responsive DonutChart so I modified your plugin to suit my need and thought this will align with other people's need as well. Do let me know, I can get some time to make that update. Do you think it will be a minor version change?

I think the quick and short  answer will be that its less complex, highly customizable and more align to the modern web development requirements. 
Although people could replicate it manually, I believe that legend is an essential feature of the chart and ability to have a responsive and customizable legend especially with css (although svg is also customizable but might be a little non-standard) and additional options to customize label with standard React Component out of the box are significant timesaver to most developers. 

Here are the preview images

Screen Shot 2023-12-28 at 12 53 04
Screen Shot 2023-12-28 at 12 53 18

**Comment by vonbearshark** [comment]
> I do agree with you that this is a radical change an update to versioning and doc is required. But I'm not sure how to approach this as I don't know what your approach is and I'm kind of busy. I needed a responsive DonutChart so I modified your plugin to suit my need and thought this will align with other people's need as well. Do let me know, I can get some time to make that update. Do you think it will be a minor version change?

Hey @ronram5126 . Again, I really appreciate your effort, but I am going to close this out. The changes generally look alright and the results look good. However, this library is hyper-focused on simplicity and stability. I think the scope of these changes would warrant a new major version, and I don't have any plans or capacity to give that the kind of attention or thinking it deserves. The intention with this library is to provide an unopinionated foundation, and given that the gist of the results could be achieved with the current functionality, I would prefer keeping the library basic

**Pull request closed without merging**